helix-view/src/editor.rs

```diff
@@ -67,6 +67,96 @@ where
     )
 }
 
+#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
+#[serde(rename_all = "kebab-case", default, deny_unknown_fields)]
+pub struct GutterConfig {
+    /// Gutter Layout
+    pub layout: Vec<GutterType>,
+    /// Options specific to the "line-numbers" gutter
+    pub line_numbers: GutterLineNumbersConfig,
+}
+
+impl Default for GutterConfig {
+    fn default() -> Self {
+        Self {
+            layout: vec![
+                GutterType::Diagnostics,
+                GutterType::Spacer,
+                GutterType::LineNumbers,
+            ],
+            line_numbers: GutterLineNumbersConfig::default(),
+        }
+    }
+}
+
+impl<'de> Deserialize<'de> for GutterConfig {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        // Accept either a simple array of strings as gutter defaults or
+        // a [editor.gutters] block
+        #[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+        #[serde(untagged)]
+        enum GutterConfigFormat {
+            Simple(Vec<GutterType>),
+            Detailed(GutterConfigStub),
+        }
+
+        // Create a mirror of a GutterConfig to let serde derive our
+        // deserialization for us.
+        // This is a hack!! Please show me a better way!!
+        #[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+        #[serde(rename_all = "kebab-case", default, deny_unknown_fields)]
+        struct GutterConfigStub {
+            layout: Vec<GutterType>,
+            line_numbers: GutterLineNumbersConfig,
+        }
```

Comment:
I'm not familiar with best practices for custom deserializers. This feels really sloppy and prone to issues with mismatches between types. Is there a better way?

Reply:
This has now been reworked and I'm much happier with how it ended up. Now the config is provided in one of two forms which has its own deserializer. When passed to a `View` it is coerced into the consistent verbose form.